### The objective behind our project is to identify the locations around which the customers can find the most number of Airbnb listings and predicting the price category of Airbnb listings namely affordable, moderate, or expensive.

In [0]:
# File location and type
file_location = "/FileStore/tables/Airbnb_project.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

host_is_superhost host_identity_verified neighbourhood_cleansed latitude longitude property_type room_type accommodates bathrooms bedrooms beds bed_type Number of amenities guests_included price_per_extra_person minimum_nights number_of_reviews number_days_btw_first_last_review review_scores_rating cancellation_policy price 0 0 Roslindale 42.2826188 -71.13306793 House Entire home/apt 4 1.5 2 3 Real Bed 14 1 0 2 0 0 null moderate 250 0 1 Roslindale 42.28624082 -71.13437396 Apartment Private room 2 1 1 1 Real Bed 20 0 0 2 36 804 94 moderate 65 1 1 Roslindale 42.29243789 -71.13576525 Apartment Private room 2 1 1 1 Real Bed 17 1 20 3 41 2574 98 moderate 65 0 0 Roslindale 42.28110619 -71.12102117 House Private room 4 1 1 2 Real Bed 22 2 25 1 1 0 100 moderate 75 1 1 Roslindale 42.28451221 -71.13625805 House Private room 2 1.5 1 2 Real Bed 13 1 0 2 29 380 99 flexible 79 1 1 Roslindale 42.2916898 -71.13189277 Condominium Private room 2 1 1 1 Real Bed 12 1 0 2 8 130 100 flexible 75 0 1 Roslindale 42.28138963 -71.13119042 Apartment Entire home/apt 3 1 1 2 Real Bed 12 1 25 1 57 421 90 strict 100 1 1 Roslindale 42.2819461 -71.14102161 House Private room 2 2 1 1 Real Bed 22 1 15 1 67 840 96 moderate 75 1 1 Roslindale 42.28587764 -71.12490956 Condominium Private room 2 1 1 2 Real Bed 9 2 0 2 65 355 96 moderate 58 1 1 Roslindale 42.28882028 -71.1395101 Apartment Entire home/apt 5 1 2 2 Real Bed 21 4 25 4 33 876 94 strict 229 0 0 Roslindale 42.2864482 -71.13932539 House Private room 2 1 1 1 Real Bed 15 1 10 1 1 0 80 flexible 60 1 1 Roslindale 42.28194183 -71.1333719 House Private room 1 1 1 1 Real Bed 13 1 0 1 17 1032 94 flexible 57 0 0 Roslindale 42.29172341 -71.13155602 Condominium Entire home/apt 4 1 2 2 Real Bed 18 1 0 1 1 0 100 flexible 93 0 1 Roslindale 42.28764103 -71.12988287 Apartment Entire home/apt 5 1 2 3 Real Bed 19 5 30 2 12 68 97 moderate 150 0 1 Roslindale 42.28626474 -71.13706513 House Entire home/apt 2 1.5 2 2 Real Bed 13 2 35 3 7 340 91 moderate 145 0 0 Roslindale 42.28145207 -71.12925004 Apartment Private room 1 0 1 1 Real Bed 14 1 0 1 70 1776 96 strict 60 0 0 Roslindale 42.27440801 -71.12529404 House Private room 4 1 1 3 Real Bed 15 3 15 3 1 0 100 flexible 165 0 0 Roslindale 42.29181492 -71.13158793 Condominium Private room 2 1 1 1 Real Bed 16 1 15 1 1 0 100 flexible 75 1 1 Roslindale 42.28222476 -71.11518856 House Private room 1 1 1 1 Real Bed 12 1 10 5 0 0 null strict 49 1 1 Roslindale 42.28280175 -71.11514509 House Private room 2 1 1 1 Real Bed 13 1 10 5 4 399 95 strict 49 1 1 Roslindale 42.27964782 -71.13605873 Villa Private room 1 2.5 1 1 Real Bed 13 1 0 2 31 742 99 flexible 40 0 1 Roslindale 42.29403534 -71.13223725 House Entire home/apt 3 1 0 2 Real Bed 17 3 0 2 16 74 98 moderate 120 0 0 Roslindale 42.28052813 -71.13084254 Apartment Private room 2 0 1 1 Real Bed 16 1 10 1 38 1794 98 strict 70 0 1 Roslindale 42.27838291 -71.1287801 House Entire home/apt 8 1 3 5 Real Bed 13 6 0 3 21 378 88 flexible 150 0 0 Roslindale 42.27683095 -71.12955974 House Entire home/apt 5 3.5 4 3 Real Bed 15 1 0 14 1 0 100 moderate 175 1 1 Roslindale 42.29090393 -71.13023771 House Private room 2 1.5 1 1 Real Bed 25 1 0 1 40 1053 96 moderate 95 0 0 Roslindale 42.29151999 -71.13185777 Apartment Private room 2 1 1 1 Real Bed 11 1 0 1 43 937 95 strict 90 1 1 Roslindale 42.28907107 -71.12823296 House Private room 2 1 1 1 Real Bed 11 1 0 2 33 231 99 moderate 95 0 1 Roslindale 42.29223824 -71.13449706 Apartment Private room 2 1 1 2 Real Bed 6 1 100 1 11 1218 90 flexible 100 1 1 Roslindale 42.28208581 -71.13334868 House Private room 2 1 1 1 Real Bed 16 1 5 1 20 1017 96 flexible 67 0 1 Roslindale 42.28456151 -71.11972634 Condominium Private room 1 1 1 1 Real Bed 15 1 0 1 30 458 92 flexible 55 0 1 Roslindale 42.27696862 -71.11775586 Apartment Entire home/apt 6 1 2 2 Real Bed 15 4 20 2 2 58 100 strict 200 0 1 Roslindale 42.28922069 -71.13071327 Apartment Entire home/apt 4 1 1 2 Real Bed 14 1 0 1 0 0 null flexible 150 1 1 Roslindale 42.28228163 -71.116

In [0]:
df.dtypes

Out[2]: [('host_is_superhost', 'string'),
 ('host_identity_verified', 'string'),
 ('neighbourhood_cleansed', 'string'),
 ('latitude', 'string'),
 ('longitude', 'string'),
 ('property_type', 'string'),
 ('room_type', 'string'),
 ('accommodates', 'string'),
 ('bathrooms', 'string'),
 ('bedrooms', 'string'),
 ('beds', 'string'),
 ('bed_type', 'string'),
 ('Number of amenities', 'string'),
 ('guests_included', 'string'),
 ('price_per_extra_person', 'string'),
 ('minimum_nights', 'string'),
 ('number_of_reviews', 'string'),
 ('number_days_btw_first_last_review', 'string'),
 ('review_scores_rating', 'string'),
 ('cancellation_policy', 'string'),
 ('price', 'string')]

In [0]:
df = df.withColumn("host_is_superhost",df.host_is_superhost.cast("int")).\
    withColumn("host_identity_verified",df.host_identity_verified.cast("int")).\
    withColumn("latitude",df.latitude.cast("double")).\
    withColumn("longitude",df.longitude.cast("double")).\
    withColumn("accommodates",df.accommodates.cast("int")).\
    withColumn("bathrooms",df.bathrooms.cast("float")).\
    withColumn("bedrooms",df.bedrooms.cast("int")).\
    withColumn("beds",df.beds.cast("int")).\
    withColumn("guests_included",df.guests_included.cast("int")).\
    withColumn("price_per_extra_person",df.price_per_extra_person.cast("int")).\
    withColumn("minimum_nights",df.minimum_nights.cast("int")).\
    withColumn("number_of_reviews",df.number_of_reviews.cast("int")).\
    withColumn("number_days_btw_first_last_review",df.number_days_btw_first_last_review.cast("int")).\
    withColumn("review_scores_rating",df.review_scores_rating.cast("int")).\
    withColumn("price",df.price.cast("int"))


In [0]:
df=df.withColumnRenamed("Number of amenities","Number_of_amenities")

In [0]:
df=df.withColumn("Number_of_amenities",df.Number_of_amenities.cast("int"))

In [0]:
df=df.withColumnRenamed("neighbourhood_cleansed","neighbourhood")

In [0]:
df.dtypes

Out[7]: [('host_is_superhost', 'int'),
 ('host_identity_verified', 'int'),
 ('neighbourhood', 'string'),
 ('latitude', 'double'),
 ('longitude', 'double'),
 ('property_type', 'string'),
 ('room_type', 'string'),
 ('accommodates', 'int'),
 ('bathrooms', 'float'),
 ('bedrooms', 'int'),
 ('beds', 'int'),
 ('bed_type', 'string'),
 ('Number_of_amenities', 'int'),
 ('guests_included', 'int'),
 ('price_per_extra_person', 'int'),
 ('minimum_nights', 'int'),
 ('number_of_reviews', 'int'),
 ('number_days_btw_first_last_review', 'int'),
 ('review_scores_rating', 'int'),
 ('cancellation_policy', 'string'),
 ('price', 'int')]

In [0]:
# Create a view or table

temp_table_name = "Airbnb_project"

df.createOrReplaceTempView(temp_table_name)

In [0]:
%sql

/* Query the created temp table in a SQL cell */

select * from `Airbnb_project`

host_is_superhost host_identity_verified neighbourhood latitude longitude property_type room_type accommodates bathrooms bedrooms beds bed_type Number_of_amenities guests_included price_per_extra_person minimum_nights number_of_reviews number_days_btw_first_last_review review_scores_rating cancellation_policy price 0 0 Roslindale 42.2826188 -71.13306793 House Entire home/apt 4 1.5 2 3 Real Bed 14 1 0 2 0 0 null moderate 250 0 1 Roslindale 42.28624082 -71.13437396 Apartment Private room 2 1.0 1 1 Real Bed 20 0 0 2 36 804 94 moderate 65 1 1 Roslindale 42.29243789 -71.13576525 Apartment Private room 2 1.0 1 1 Real Bed 17 1 20 3 41 2574 98 moderate 65 0 0 Roslindale 42.28110619 -71.12102117 House Private room 4 1.0 1 2 Real Bed 22 2 25 1 1 0 100 moderate 75 1 1 Roslindale 42.28451221 -71.13625805 House Private room 2 1.5 1 2 Real Bed 13 1 0 2 29 380 99 flexible 79 1 1 Roslindale 42.2916898 -71.13189277 Condominium Private room 2 1.0 1 1 Real Bed 12 1 0 2 8 130 100 flexible 75 0 1 Roslindale 42.28138963 -71.13119042 Apartment Entire home/apt 3 1.0 1 2 Real Bed 12 1 25 1 57 421 90 strict 100 1 1 Roslindale 42.2819461 -71.14102161 House Private room 2 2.0 1 1 Real Bed 22 1 15 1 67 840 96 moderate 75 1 1 Roslindale 42.28587764 -71.12490956 Condominium Private room 2 1.0 1 2 Real Bed 9 2 0 2 65 355 96 moderate 58 1 1 Roslindale 42.28882028 -71.1395101 Apartment Entire home/apt 5 1.0 2 2 Real Bed 21 4 25 4 33 876 94 strict 229 0 0 Roslindale 42.2864482 -71.13932539 House Private room 2 1.0 1 1 Real Bed 15 1 10 1 1 0 80 flexible 60 1 1 Roslindale 42.28194183 -71.1333719 House Private room 1 1.0 1 1 Real Bed 13 1 0 1 17 1032 94 flexible 57 0 0 Roslindale 42.29172341 -71.13155602 Condominium Entire home/apt 4 1.0 2 2 Real Bed 18 1 0 1 1 0 100 flexible 93 0 1 Roslindale 42.28764103 -71.12988287 Apartment Entire home/apt 5 1.0 2 3 Real Bed 19 5 30 2 12 68 97 moderate 150 0 1 Roslindale 42.28626474 -71.13706513 House Entire home/apt 2 1.5 2 2 Real Bed 13 2 35 3 7 340 91 moderate 145 0 0 Roslindale 42.28145207 -71.12925004 Apartment Private room 1 0.0 1 1 Real Bed 14 1 0 1 70 1776 96 strict 60 0 0 Roslindale 42.27440801 -71.12529404 House Private room 4 1.0 1 3 Real Bed 15 3 15 3 1 0 100 flexible 165 0 0 Roslindale 42.29181492 -71.13158793 Condominium Private room 2 1.0 1 1 Real Bed 16 1 15 1 1 0 100 flexible 75 1 1 Roslindale 42.28222476 -71.11518856 House Private room 1 1.0 1 1 Real Bed 12 1 10 5 0 0 null strict 49 1 1 Roslindale 42.28280175 -71.11514509 House Private room 2 1.0 1 1 Real Bed 13 1 10 5 4 399 95 strict 49 1 1 Roslindale 42.27964782 -71.13605873 Villa Private room 1 2.5 1 1 Real Bed 13 1 0 2 31 742 99 flexible 40 0 1 Roslindale 42.29403534 -71.13223725 House Entire home/apt 3 1.0 0 2 Real Bed 17 3 0 2 16 74 98 moderate 120 0 0 Roslindale 42.28052813 -71.13084254 Apartment Private room 2 0.0 1 1 Real Bed 16 1 10 1 38 1794 98 strict 70 0 1 Roslindale 42.27838291 -71.1287801 House Entire home/apt 8 1.0 3 5 Real Bed 13 6 0 3 21 378 88 flexible 150 0 0 Roslindale 42.27683095 -71.12955974 House Entire home/apt 5 3.5 4 3 Real Bed 15 1 0 14 1 0 100 moderate 175 1 1 Roslindale 42.29090393 -71.13023771 House Private room 2 1.5 1 1 Real Bed 25 1 0 1 40 1053 96 moderate 95 0 0 Roslindale 42.29151999 -71.13185777 Apartment Private room 2 1.0 1 1 Real Bed 11 1 0 1 43 937 95 strict 90 1 1 Roslindale 42.28907107 -71.12823296 House Private room 2 1.0 1 1 Real Bed 11 1 0 2 33 231 99 moderate 95 0 1 Roslindale 42.29223824 -71.13449706 Apartment Private room 2 1.0 1 2 Real Bed 6 1 100 1 11 1218 90 flexible 100 1 1 Roslindale 42.28208581 -71.13334868 House Private room 2 1.0 1 1 Real Bed 16 1 5 1 20 1017 96 flexible 67 0 1 Roslindale 42.28456151 -71.11972634 Condominium Private room 1 1.0 1 1 Real Bed 15 1 0 1 30 458 92 flexible 55 0 1 Roslindale 42.27696862 -71.11775586 Apartment Entire home/apt 6 1.0 2 2 Real Bed 15 4 20 2 2 58 100 strict 200 0 1 Roslindale 42.28922069 -71.13071327 Apartment Entire home/apt 4 1.0 1 2 Real Bed 14 1 0 1 0 0 null fl

In [0]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.

permanent_table_name = "Airbnbproject"

# df.write.format("parquet").mode("overwrite").saveAsTable(permanent_table_name)

In [0]:
df=df.dropna()

### Which neighborhood has maximum number of listings?

In [0]:
from pyspark.sql.functions import *

df.groupBy("neighbourhood").count().sort(desc('count')).show()

+--------------------+-----+
|       neighbourhood|count|
+--------------------+-----+
|       Jamaica Plain|  837|
|           South End|  750|
|            Back Bay|  678|
|          Dorchester|  657|
|              Fenway|  546|
|             Allston|  540|
|         Beacon Hill|  477|
|        South Boston|  423|
|         East Boston|  396|
|            Brighton|  390|
|            Downtown|  339|
|           North End|  339|
|             Roxbury|  327|
|        Mission Hill|  243|
|         Charlestown|  207|
|South Boston Wate...|  153|
|          Roslindale|  150|
|           Chinatown|  123|
|        West Roxbury|   99|
|            West End|   81|
+--------------------+-----+
only showing top 20 rows



### Average price for each property type

In [0]:
%sql
select avg(CASE WHEN property_type='Apartment' THEN price else null end) as avg_apartment_price, 
       avg(CASE WHEN property_type='Condominium' THEN price else null end) as avg_condo_price,
       avg(CASE WHEN property_type='House' THEN price else null end) as avg_house_price,
       avg(CASE WHEN property_type='Villa' THEN price else null end) as avg_villa_price,
       avg(CASE WHEN property_type='Loft' THEN price else null end) as avg_loft_price,
       avg(CASE WHEN property_type='Townhouse' THEN price else null end) as avg_townhouse_price
from Airbnb_project

avg_apartment_price avg_condo_price avg_house_price avg_villa_price avg_loft_price avg_townhouse_price 168.34115805946792 172.97297297297297 113.69981583793738 189.5 177.1315789473684 123.0

### Neighbourhood with highest average rating (top 5)

In [0]:
df.groupby("neighbourhood").agg(avg("Review_scores_rating").alias('Avg_rating')).sort(desc('Avg_rating')).show(5)

+--------------------+-----------------+
|       neighbourhood|       Avg_rating|
+--------------------+-----------------+
|    Leather District|98.33333333333333|
|          Roslindale|            95.38|
|        West Roxbury|95.21212121212122|
|South Boston Wate...|94.45098039215686|
|        South Boston|94.39716312056737|
+--------------------+-----------------+
only showing top 5 rows



### Neighborhood with highest average price

In [0]:
df.groupby("neighbourhood").avg("price").sort(desc("avg(price)")).show(1)

+----------------+------------------+
|   neighbourhood|        avg(price)|
+----------------+------------------+
|Leather District|242.66666666666666|
+----------------+------------------+
only showing top 1 row



### Property type percentage in the overall region

In [0]:
from pyspark.sql.window import Window

df1=df.groupby(["neighbourhood","property_type"]).count().\
    withColumn("percentage",round(col("count")*100/sum("count").over(Window.partitionBy("neighbourhood")),2))
display(df1.sort("neighbourhood",desc("percentage")))

neighbourhood property_type count percentage Allston Apartment 357 66.11 Allston House 159 29.44 Allston Condominium 21 3.89 Allston Townhouse 3 0.56 Back Bay Apartment 618 91.15 Back Bay Condominium 42 6.19 Back Bay Loft 9 1.33 Back Bay Townhouse 6 0.88 Back Bay House 3 0.44 Bay Village Apartment 27 81.82 Bay Village Guesthouse 3 9.09 Bay Village Condominium 3 9.09 Beacon Hill Apartment 438 91.82 Beacon Hill Condominium 30 6.29 Beacon Hill House 6 1.26 Beacon Hill Townhouse 3 0.63 Brighton Apartment 255 65.38 Brighton House 111 28.46 Brighton Condominium 12 3.08 Brighton Townhouse 9 2.31 Brighton Loft 3 0.77 Charlestown Apartment 138 66.67 Charlestown House 39 18.84 Charlestown Condominium 27 13.04 Charlestown Entire Floor 3 1.45 Chinatown Apartment 108 87.8 Chinatown Loft 12 9.76 Chinatown Condominium 3 2.44 Dorchester Apartment 390 59.36 Dorchester House 216 32.88 Dorchester Condominium 36 5.48 Dorchester Townhouse 15 2.28 Downtown Apartment 306 90.27 Downtown Loft 15 4.42 Downtown Condominium 9 2.65 Downtown House 6 1.77 Downtown Villa 3 0.88 East Boston Apartment 297 75.0 East Boston House 66 16.67 East Boston Condominium 27 6.82 East Boston Townhouse 6 1.52 Fenway Apartment 531 97.25 Fenway Condominium 9 1.65 Fenway House 6 1.1 Hyde Park House 48 66.67 Hyde Park Apartment 18 25.0 Hyde Park Condominium 3 4.17 Hyde Park Loft 3 4.17 Jamaica Plain Apartment 492 58.78 Jamaica Plain House 249 29.75 Jamaica Plain Condominium 57 6.81 Jamaica Plain Townhouse 33 3.94 Jamaica Plain Loft 6 0.72 Leather District Condominium 6 66.67 Leather District Apartment 3 33.33 Longwood Medical Area Apartment 6 66.67 Longwood Medical Area Condominium 3 33.33 Mattapan House 39 81.25 Mattapan Apartment 9 18.75 Mission Hill Apartment 183 75.31 Mission Hill House 36 14.81 Mission Hill Loft 12 4.94 Mission Hill Condominium 9 3.7 Mission Hill Townhouse 3 1.23 North End Apartment 297 87.61 North End Condominium 18 5.31 North End House 12 3.54 North End Loft 9 2.65 North End Townhouse 3 0.88 Roslindale House 90 60.0 Roslindale Apartment 42 28.0 Roslindale Condominium 15 10.0 Roslindale Villa 3 2.0 Roxbury Apartment 216 66.06 Roxbury House 96 29.36 Roxbury Condominium 9 2.75 Roxbury Townhouse 6 1.83 South Boston Apartment 225 53.19 South Boston House 99 23.4 South Boston Condominium 78 18.44 South Boston Townhouse 18 4.26 South Boston Loft 3 0.71 South Boston Waterfront Apartment 129 84.31 South Boston Waterfront Loft 12 7.84 South Boston Waterfront Condominium 9 5.88 South Boston Waterfront House 3 1.96 South End Apartment 624 83.2 South End Condominium 81 10.8 South End House 18 2.4 South End Townhouse 18 2.4 South End Loft 9 1.2 West End Apartment 78 96.3 West End Condominium 3 3.7 West Roxbury House 69 69.7 West Roxbury Apartment 24 24.24 West Roxbury Townhouse 6 6.06

Output can only be rendered in Databricks

### Creating new column for categorizing price

In [0]:
display(df.select("price")) ### Creating a histogram for categorizing the price

price 65 65 75 79 75 100 75 58 229 60 57 93 150 145 60 165 75 49 40 120 70 150 175 95 90 95 100 67 55 200 49 75 110 125 200 75 78 50 60 45 145 45 85 75 50 45 60 95 89 285 70 95 80 120 65 100 65 63 75 65 110 46 62 69 73 85 120 125 70 61 125 100 80 85 75 57 100 70 190 149 168 205 65 75 429 95 69 165 150 87 62 110 250 150 300 119 175 195 145 200 90 210 75 46 75 60 130 190 100 450 399 350 200 99 107 69 76 75 55 65 88 90 185 105 90 100 100 200 300 319 85 70 50 119 125 53 125 199 205 114 75 185 75 70 120 92 84 200 115 95 275 230 130 192 175 70 65 100 250 110 168 60 59 60 145 80 114 170 240 150 69 80 85 75 150 70 80 225 178 175 109 383 99 50 300 80 85 150 100 100 84 130 76 100 70 125 70 170 95 250 115 70 135 120 130 29 68 300 50 119 46 130 180 136 100 88 95 135 135 189 75 55 125 88 72 22 90 82 84 160 110 150 80 225 65 275 300 185 175 75 60 100 75 94 85 83 75 168 60 135 130 65 77 71 200 55 220 72 70 135 300 110 70 75 70 480 180 325 290 159 200 95 65 99 93 295 450 150 150 399 71 61 250 56 200 220 125 149 65 99 67 180 170 115 89 110 84 65 86 75 255 145 199 91 81 60 99 88 225 50 225 48 85 100 75 25 50 49 60 85 85 64 75 119 110 179 175 125 220 239 75 30 45 160 55 79 79 75 105 85 110 199 45 269 69 45 70 70 99 90 80 275 71 125 52 70 208 399 100 75 115 65 75 199 160 65 41 49 35 72 90 100 170 169 150 56 140 88 74 40 80 50 103 42 169 89 150 200 99 272 85 70 109 79 28 55 110 40 90 74 100 130 246 150 45 100 130 60 246 84 149 84 165 289 139 165 350 188 400 180 125 90 225 390 159 179 399 219 295 259 185 269 109 295 389 375 285 296 219 120 399 159 125 80 299 100 383 149 250 349 300 169 179 319 135 169 95 287 95 389 169 255 299 199 93 299 349 140 159 84 125 150 79 320 180 150 125 145 125 285 400 400 210 350 182 200 200 150 120 210 159 120 150 149 115 150 220 110 125 310 160 175 159 165 195 195 300 135 125 125 425 180 165 120 129 160 145 131 150 200 200 125 250 250 109 180 249 135 300 103 120 250 69 125 240 180 60 200 380 220 180 140 200 239 199 175 175 225 325 175 179 99 165 290 179 95 152 200 130 175 139 255 119 135 200 275 139 199 135 72 200 175 175 200 195 235 195 150 235 185 35 139 145 79 209 99 90 79 22 41 145 132 167 75 80 265 150 59 165 195 275 225 109 85 75 65 159 95 65 99 135 100 65 99 135 89 100 75 90 94 125 99 100 92 100 170 33 52 78 350 236 330 300 60 150 83 55 120 450 90 350 179 130 60 110 75 120 60 55 195 114 150 39 95 450 269 250 120 229 175 70 150 75 229 450 99 99 92 125 85 35 90 35 65 40 50 200 90 65 100 35 63 63 167 150 150 225 81 79 195 295 250 150 279 279 290 185 399 255 75 250 100 215 125 110 125 80 373 299 269 189 215 279 289 229 154 236 175 229 234 329 209 195 169 145 219 150 215 155 138 170 195 125 250 100 137 199 239 185 175 200 175 125 150 200 275 125 265 160 155 99 125 235 149 100 164 185 250 275 225 250 115 150 170 125 250 229 90 200 100 306 229 199 67 120 239 197 85 266 160 280 225 170 183 300 166 350 190 180 99 110 210 155 178 170 129 160 149 175 150 250 100 200 120 150 152 200 170 157 229 210 110 200 115 110 250 200 417 175 69 150 133 299 135 275 99 170 450 200 189 150 179 250 229 250 89 80 140 84 200 150 119 300 170 170 100 159 250 225 75 144 170 120 139 147 217 200 150 279 135 240 77 160 175 228 145 89 238 161 149 300 177 185 240 225 125 150 122 170 160 199 160 200 200 350 139 80 215 130 183 188 215 75 125 290 215 255 300 250 155 200 119 189 250 254 137 146 199 150 200 99 139 175 200 450 165 350 295 199 130 160 220 150 280 150 250 185 140 150 165 275 225 235 189 90 250 195 225 149 379 359 299 110 170 130 259 175 289 80 251 185 251 349 389 342 232 249 135 280 155 200 150 239 174 199 425 160 199 150 150 150 148 160 300 160 100 225 363 105 185 175 175 80 219 350 232 180 189 165 375 298 245 363 259 276 359 165 135 107 150 Showing the first 1000 rows.

Output can only be rendered in Databricks

### We will categorize price into 3 categories
1. Affordable (less than $175)
2. Moderate (between $175 and $335)
3. Expensive (greater than $335)

In [0]:
df2=df.select(['host_is_superhost','host_identity_verified','neighbourhood','latitude','longitude','property_type','room_type','accommodates','bathrooms','bedrooms','beds','bed_type','Number_of_amenities','guests_included','price_per_extra_person','minimum_nights','number_of_reviews','number_days_btw_first_last_review','review_scores_rating','cancellation_policy','price'])

In [0]:
def price_categorizer(x):
    price_c=''
    if x.price <= 175:
        price_c = 'Affordable'
    elif x.price > 175 and x.price <=335:
        price_c = 'Moderate'
    else:
        price_c= 'Expensive'
    return(price_c,x.price)
rdd2 = df2.rdd.map(lambda x : price_categorizer(x))

In [0]:
df4=rdd2.toDF(["price_c","price2"])

In [0]:
data = df2.join(df4,df2.price == df4.price2,"inner")

In [0]:
data=data.drop("price2")
display(data)

host_is_superhost host_identity_verified neighbourhood latitude longitude property_type room_type accommodates bathrooms bedrooms beds bed_type Number_of_amenities guests_included price_per_extra_person minimum_nights number_of_reviews number_days_btw_first_last_review review_scores_rating cancellation_policy price price_c 0 1 Charlestown 42.38522635 -71.08092263 Apartment Private room 2 1.0 1 1 Real Bed 7 1 30 1 2 8 90 strict 65 Affordable 0 1 Charlestown 42.38774101 -71.07989661 Apartment Private room 2 1.0 1 1 Real Bed 4 1 30 1 2 12 80 flexible 65 Affordable 0 1 Allston 42.34832593 -71.13613371 House Entire home/apt 3 1.0 1 2 Real Bed 10 1 0 1 1 0 80 flexible 65 Affordable 0 0 Allston 42.34877625 -71.13132203 Apartment Entire home/apt 2 1.0 1 1 Futon 13 1 0 3 9 64 77 flexible 65 Affordable 0 1 Allston 42.3533978 -71.14235466 House Private room 2 2.0 1 1 Real Bed 18 1 20 30 36 591 87 strict 65 Affordable 1 1 Allston 42.34998843 -71.14420208 House Private room 2 2.0 1 1 Real Bed 12 1 20 1 39 713 96 strict 65 Affordable 0 0 South Boston 42.33084495 -71.05246031 House Private room 2 1.0 1 1 Real Bed 13 0 0 2 11 67 98 moderate 65 Affordable 0 0 Dorchester 42.31673406 -71.05337817 House Private room 3 1.0 1 1 Real Bed 4 1 0 1 14 66 80 flexible 65 Affordable 0 1 Dorchester 42.2924911 -71.07464279 House Private room 2 1.0 1 1 Real Bed 7 1 0 2 15 456 73 strict 65 Affordable 0 1 Dorchester 42.31417775 -71.060367 House Private room 2 2.5 1 1 Real Bed 14 1 20 2 62 782 91 flexible 65 Affordable 0 1 Dorchester 42.28373701 -71.08581156 House Private room 2 1.0 1 1 Airbed 18 1 20 1 14 139 90 strict 65 Affordable 0 1 Dorchester 42.31873349 -71.0536317 Apartment Private room 1 1.0 1 1 Real Bed 11 1 20 1 7 386 89 flexible 65 Affordable 0 1 Dorchester 42.31797689 -71.05501438 Apartment Private room 2 1.0 1 1 Real Bed 7 1 0 4 16 731 70 strict 65 Affordable 0 1 Dorchester 42.28439248 -71.06911628 Apartment Private room 2 1.0 1 1 Real Bed 22 2 40 4 43 816 93 flexible 65 Affordable 0 1 Dorchester 42.31189661 -71.07336743 Condominium Private room 2 1.0 1 1 Real Bed 18 1 25 2 3 43 93 flexible 65 Affordable 0 1 Dorchester 42.29849091 -71.05880244 House Private room 2 1.5 1 1 Real Bed 14 1 15 2 125 1806 89 strict 65 Affordable 0 1 Dorchester 42.29887615 -71.06489096 House Private room 2 1.5 1 2 Real Bed 6 1 0 1 120 1190 89 moderate 65 Affordable 0 1 Dorchester 42.31339235 -71.05665201 Apartment Private room 2 1.0 1 1 Real Bed 4 2 0 3 4 376 55 strict 65 Affordable 0 1 Dorchester 42.29295129 -71.07317264 House Private room 2 1.0 1 1 Real Bed 10 1 0 2 23 488 74 strict 65 Affordable 0 1 Dorchester 42.31998169 -71.05439069 Apartment Private room 2 1.0 1 1 Real Bed 5 1 0 4 2 32 50 strict 65 Affordable 0 1 Dorchester 42.29620501 -71.04894322 Apartment Private room 2 1.0 1 1 Real Bed 14 1 0 1 13 60 94 flexible 65 Affordable 0 1 Dorchester 42.29977341 -71.06047581 House Private room 4 2.5 1 2 Real Bed 15 1 15 1 291 2260 87 strict 65 Affordable 0 1 Dorchester 42.2910007 -71.07480272 House Private room 2 1.0 1 1 Real Bed 7 1 0 2 16 451 71 strict 65 Affordable 0 1 Dorchester 42.29770901 -71.05979096 House Private room 3 1.5 1 1 Real Bed 18 1 15 1 194 1991 88 strict 65 Affordable 0 1 Dorchester 42.28355381 -71.08444105 House Private room 1 1.0 1 2 Real Bed 18 1 20 1 4 112 80 strict 65 Affordable 0 1 Dorchester 42.28284157 -71.07115571 Apartment Private room 2 1.0 1 1 Real Bed 22 2 40 4 40 822 93 flexible 65 Affordable 0 1 Mattapan 42.28189243 -71.09362734 Apartment Entire home/apt 5 1.0 2 2 Real Bed 16 3 35 4 10 130 84 strict 65 Affordable 0 1 Mattapan 42.27513775 -71.07739962 House Private room 2 1.5 1 2 Real Bed 13 1 40 1 9 19 100 strict 65 Affordable 0 0 Hyde Park 42.26653249 -71.10883175 House Private room 2 1.5 1 1 Real Bed 19 1 25 1 1 0 100 strict 65 Affordable 1 0 West Roxbury 42.2724567 -71.16863145 House Private room 3 3.5 1 2 Real Bed 7 1 20 1 22 103 95 strict 65 Affordable 1 0 West Roxbury 42.27426067 -71.16704389 House Private room 4 3.5 1 2 Real Bed 1

### K means clustering

In [0]:
model_1= data.select(["latitude","longitude"])

#### Creating Elbow Plot for determining optimal number of clusters

In [0]:
display(model_1)

latitude longitude 42.38522635 -71.08092263 42.38774101 -71.07989661 42.34832593 -71.13613371 42.34877625 -71.13132203 42.3533978 -71.14235466 42.34998843 -71.14420208 42.33084495 -71.05246031 42.31673406 -71.05337817 42.2924911 -71.07464279 42.31417775 -71.060367 42.28373701 -71.08581156 42.31873349 -71.0536317 42.31797689 -71.05501438 42.28439248 -71.06911628 42.31189661 -71.07336743 42.29849091 -71.05880244 42.29887615 -71.06489096 42.31339235 -71.05665201 42.29295129 -71.07317264 42.31998169 -71.05439069 42.29620501 -71.04894322 42.29977341 -71.06047581 42.2910007 -71.07480272 42.29770901 -71.05979096 42.28355381 -71.08444105 42.28284157 -71.07115571 42.28189243 -71.09362734 42.27513775 -71.07739962 42.26653249 -71.10883175 42.2724567 -71.16863145 42.27426067 -71.16704389 42.27408555 -71.168792 42.29419543 -71.13817106 42.34916139 -71.16347987 42.34985003 -71.16451315 42.33588717 -71.14895443 42.35057716 -71.1639392 42.34488949 -71.13714359 42.34532141 -71.0874855 42.36622622 -71.03143059 42.37240877 -71.03867839 42.36959472 -71.0399765 42.37664361 -71.03147889 42.36664961 -71.03786795 42.36699756 -71.03594265 42.36937694 -71.03334518 42.32554505 -71.08577829 42.33720962 -71.08478305 42.33588188 -71.08310262 42.33539829 -71.08259403 42.31643567 -71.09391225 42.32739596 -71.10508282 42.33320209 -71.10146471 42.29661804 -71.11369881 42.32136495 -71.11323106 42.32002036 -71.10745885 42.32443218 -71.10789033 42.30601751 -71.10953839 42.32235541 -71.10672991 42.32249609 -71.11475792 42.31362383 -71.09821201 42.31549953 -71.11706562 42.32171967 -71.10345882 42.32002252 -71.1059405 42.29243789 -71.13576525 42.28624082 -71.13437396 42.38522635 -71.08092263 42.38774101 -71.07989661 42.34832593 -71.13613371 42.34877625 -71.13132203 42.3533978 -71.14235466 42.34998843 -71.14420208 42.33084495 -71.05246031 42.31673406 -71.05337817 42.2924911 -71.07464279 42.31417775 -71.060367 42.28373701 -71.08581156 42.31873349 -71.0536317 42.31797689 -71.05501438 42.28439248 -71.06911628 42.31189661 -71.07336743 42.29849091 -71.05880244 42.29887615 -71.06489096 42.31339235 -71.05665201 42.29295129 -71.07317264 42.31998169 -71.05439069 42.29620501 -71.04894322 42.29977341 -71.06047581 42.2910007 -71.07480272 42.29770901 -71.05979096 42.28355381 -71.08444105 42.28284157 -71.07115571 42.28189243 -71.09362734 42.27513775 -71.07739962 42.26653249 -71.10883175 42.2724567 -71.16863145 42.27426067 -71.16704389 42.27408555 -71.168792 42.29419543 -71.13817106 42.34916139 -71.16347987 42.34985003 -71.16451315 42.33588717 -71.14895443 42.35057716 -71.1639392 42.34488949 -71.13714359 42.34532141 -71.0874855 42.36622622 -71.03143059 42.37240877 -71.03867839 42.36959472 -71.0399765 42.37664361 -71.03147889 42.36664961 -71.03786795 42.36699756 -71.03594265 42.36937694 -71.03334518 42.32554505 -71.08577829 42.33720962 -71.08478305 42.33588188 -71.08310262 42.33539829 -71.08259403 42.31643567 -71.09391225 42.32739596 -71.10508282 42.33320209 -71.10146471 42.29661804 -71.11369881 42.32136495 -71.11323106 42.32002036 -71.10745885 42.32443218 -71.10789033 42.30601751 -71.10953839 42.32235541 -71.10672991 42.32249609 -71.11475792 42.31362383 -71.09821201 42.31549953 -71.11706562 42.32171967 -71.10345882 42.32002252 -71.1059405 42.29243789 -71.13576525 42.28624082 -71.13437396 42.38522635 -71.08092263 42.38774101 -71.07989661 42.34832593 -71.13613371 42.34877625 -71.13132203 42.3533978 -71.14235466 42.34998843 -71.14420208 42.33084495 -71.05246031 42.31673406 -71.05337817 42.2924911 -71.07464279 42.31417775 -71.060367 42.28373701 -71.08581156 42.31873349 -71.0536317 42.31797689 -71.05501438 42.28439248 -71.06911628 42.31189661 -71.07336743 42.29849091 -71.05880244 42.29887615 -71.06489096 42.31339235 -71.05665201 42.29295129 -71.07317264 42.31998169 -71.05439069 42.29620501 -71.04894322 42.29977341 -71.06047581 42.2910007 -71.07480272 42.29770901 -71.05979096 42.28355381 -71.08444105 42.28284157 -71.07115571 42.28189243 -71.09362734 42.27513775 -71.07739962 42.26653249 -

In [0]:
from pyspark.ml.clustering import KMeans
from pyspark.ml.feature import VectorAssembler,StringIndexer
from pyspark.ml import Pipeline

features = ["latitude","longitude"]
assembler = VectorAssembler(
    inputCols=features, 
    outputCol="airbnb_features")
assembled_data = assembler.transform(model_1)


In [0]:
costs = []
for k in range(2, 10):
    k_means = KMeans(featuresCol='airbnb_features', k=k)
    model = k_means.fit(assembled_data)
    costs.append((k,model.summary.trainingCost))

In [0]:
# import pylab as pl
# pl.plot(costs.keys(), costs.values())
# pl.xlabel('Number of Clusters')
# pl.ylabel('Score')
# pl.title('Elbow Curve')
# pl.show()

In [0]:
elbow_df=spark.createDataFrame(costs,schema=["Number_of_clusters","Score"])

In [0]:
display(elbow_df)

Number_of_clusters Score 2 738.9625484686453 3 540.8303627714592 4 429.2028074874493 5 290.2114114993138 6 232.23400213932163 7 191.99683530061026 8 142.64928877137388 9 145.84368307872492

Output can only be rendered in Databricks

In [0]:
assembler = VectorAssembler(inputCols=['latitude','longitude'],
                            outputCol="features")

In [0]:
pipe = Pipeline(stages=[assembler])

In [0]:
result_model1=pipe.fit(model_1).transform(model_1)

In [0]:
kmeans_model = KMeans(k=5)

In [0]:
fit_model = kmeans_model.fit(result_model1)

In [0]:
# wsse = fit_model.computeCost(final_data) for spark 2.7
wssse = fit_model.summary.trainingCost # for spark 3.0
print("The within set sum of squared error of the mode is {}".format(wssse))

The within set sum of squared error of the mode is 290.2114114993138


In [0]:
centers = fit_model.clusterCenters()

In [0]:
print("Cluster Centers")
index=1
for cluster in centers:
    print("Centroid {}: {}".format(index,cluster))
    index+=1
#'opened_by_index','location_index','category_index',
# 'subcategory_index','u_symptom_index','assignment_group_index'

Cluster Centers
Centroid 1: [ 42.31555492 -71.05618115]
Centroid 2: [ 42.36325513 -71.05463656]
Centroid 3: [ 42.35114839 -71.13995305]
Centroid 4: [ 42.34186255 -71.08475597]
Centroid 5: [ 42.30114796 -71.11702549]


In [0]:
results = fit_model.transform(result_model1)

In [0]:
display(results.select(['latitude','longitude','prediction']))

latitude longitude prediction 42.38522635 -71.08092263 1 42.38774101 -71.07989661 1 42.34832593 -71.13613371 2 42.34877625 -71.13132203 2 42.3533978 -71.14235466 2 42.34998843 -71.14420208 2 42.33084495 -71.05246031 0 42.31673406 -71.05337817 0 42.2924911 -71.07464279 0 42.31417775 -71.060367 0 42.28373701 -71.08581156 4 42.31873349 -71.0536317 0 42.31797689 -71.05501438 0 42.28439248 -71.06911628 0 42.31189661 -71.07336743 0 42.29849091 -71.05880244 0 42.29887615 -71.06489096 0 42.31339235 -71.05665201 0 42.29295129 -71.07317264 0 42.31998169 -71.05439069 0 42.29620501 -71.04894322 0 42.29977341 -71.06047581 0 42.2910007 -71.07480272 0 42.29770901 -71.05979096 0 42.28355381 -71.08444105 4 42.28284157 -71.07115571 0 42.28189243 -71.09362734 4 42.27513775 -71.07739962 0 42.26653249 -71.10883175 4 42.2724567 -71.16863145 4 42.27426067 -71.16704389 4 42.27408555 -71.168792 4 42.29419543 -71.13817106 4 42.34916139 -71.16347987 2 42.34985003 -71.16451315 2 42.33588717 -71.14895443 2 42.35057716 -71.1639392 2 42.34488949 -71.13714359 2 42.34532141 -71.0874855 3 42.36622622 -71.03143059 1 42.37240877 -71.03867839 1 42.36959472 -71.0399765 1 42.37664361 -71.03147889 1 42.36664961 -71.03786795 1 42.36699756 -71.03594265 1 42.36937694 -71.03334518 1 42.32554505 -71.08577829 3 42.33720962 -71.08478305 3 42.33588188 -71.08310262 3 42.33539829 -71.08259403 3 42.31643567 -71.09391225 3 42.32739596 -71.10508282 3 42.33320209 -71.10146471 3 42.29661804 -71.11369881 4 42.32136495 -71.11323106 4 42.32002036 -71.10745885 4 42.32443218 -71.10789033 4 42.30601751 -71.10953839 4 42.32235541 -71.10672991 4 42.32249609 -71.11475792 4 42.31362383 -71.09821201 4 42.31549953 -71.11706562 4 42.32171967 -71.10345882 4 42.32002252 -71.1059405 4 42.29243789 -71.13576525 4 42.28624082 -71.13437396 4 42.38522635 -71.08092263 1 42.38774101 -71.07989661 1 42.34832593 -71.13613371 2 42.34877625 -71.13132203 2 42.3533978 -71.14235466 2 42.34998843 -71.14420208 2 42.33084495 -71.05246031 0 42.31673406 -71.05337817 0 42.2924911 -71.07464279 0 42.31417775 -71.060367 0 42.28373701 -71.08581156 4 42.31873349 -71.0536317 0 42.31797689 -71.05501438 0 42.28439248 -71.06911628 0 42.31189661 -71.07336743 0 42.29849091 -71.05880244 0 42.29887615 -71.06489096 0 42.31339235 -71.05665201 0 42.29295129 -71.07317264 0 42.31998169 -71.05439069 0 42.29620501 -71.04894322 0 42.29977341 -71.06047581 0 42.2910007 -71.07480272 0 42.29770901 -71.05979096 0 42.28355381 -71.08444105 4 42.28284157 -71.07115571 0 42.28189243 -71.09362734 4 42.27513775 -71.07739962 0 42.26653249 -71.10883175 4 42.2724567 -71.16863145 4 42.27426067 -71.16704389 4 42.27408555 -71.168792 4 42.29419543 -71.13817106 4 42.34916139 -71.16347987 2 42.34985003 -71.16451315 2 42.33588717 -71.14895443 2 42.35057716 -71.1639392 2 42.34488949 -71.13714359 2 42.34532141 -71.0874855 3 42.36622622 -71.03143059 1 42.37240877 -71.03867839 1 42.36959472 -71.0399765 1 42.37664361 -71.03147889 1 42.36664961 -71.03786795 1 42.36699756 -71.03594265 1 42.36937694 -71.03334518 1 42.32554505 -71.08577829 3 42.33720962 -71.08478305 3 42.33588188 -71.08310262 3 42.33539829 -71.08259403 3 42.31643567 -71.09391225 3 42.32739596 -71.10508282 3 42.33320209 -71.10146471 3 42.29661804 -71.11369881 4 42.32136495 -71.11323106 4 42.32002036 -71.10745885 4 42.32443218 -71.10789033 4 42.30601751 -71.10953839 4 42.32235541 -71.10672991 4 42.32249609 -71.11475792 4 42.31362383 -71.09821201 4 42.31549953 -71.11706562 4 42.32171967 -71.10345882 4 42.32002252 -71.1059405 4 42.29243789 -71.13576525 4 42.28624082 -71.13437396 4 42.38522635 -71.08092263 1 42.38774101 -71.07989661 1 42.34832593 -71.13613371 2 42.34877625 -71.13132203 2 42.3533978 -71.14235466 2 42.34998843 -71.14420208 2 42.33084495 -71.05246031 0 42.31673406 -71.05337817 0 42.2924911 -71.07464279 0 42.31417775 -71.060367 0 42.28373701 -71.08581156 4 42.31873349 -71.0536317 0 42.31797689 -71.05501438 0 42.28439248 -71.06911628 0 42.31189661 -71.07336743 0 42.29849091 -71.05880244 0 42.29

In [0]:
display(results.groupby(['prediction']).count().sort(desc('count')))

prediction count 3 253629 1 242766 4 140193 0 123426 2 96246

Output can only be rendered in Databricks

### Interpretation

Most of the listings(about 58%) belong to cluster 3 and 1 which means that:

29.6 % of the listings are centered around location with latitude 42.35114839 and longitude -71.13995305, which are the co-ordinates for <B>50 Gordon St, Boston, MA area.</B>

28.6 % of the listings are centered around location with latitude 42.31555492 and longitude -71.05618115, which are the co-ordinates for <B>68 Belfort St, Dorchester, Boston, MA area.</B>

It can be suggested that tourists have higher chances of finding Airbnb listings for stay around the above suggested areas.

### Classification Model 1 - Decision Tree Classification

In [0]:
#data_model2= data.select(['host_is_superhost','neighbourhood','property_type','room_type','accommodates','bathrooms','bedrooms','Number_of_amenities','price_c'])

data_model2= data.select(['host_is_superhost','host_identity_verified','neighbourhood','property_type','room_type','accommodates','bathrooms','bedrooms','beds','bed_type','Number_of_amenities','guests_included','price_per_extra_person','minimum_nights','number_of_reviews','number_days_btw_first_last_review','cancellation_policy','price_c'])



In [0]:
data_model2.dtypes

Out[42]: [('host_is_superhost', 'int'),
 ('host_identity_verified', 'int'),
 ('neighbourhood', 'string'),
 ('property_type', 'string'),
 ('room_type', 'string'),
 ('accommodates', 'int'),
 ('bathrooms', 'float'),
 ('bedrooms', 'int'),
 ('beds', 'int'),
 ('bed_type', 'string'),
 ('Number_of_amenities', 'int'),
 ('guests_included', 'int'),
 ('price_per_extra_person', 'int'),
 ('minimum_nights', 'int'),
 ('number_of_reviews', 'int'),
 ('number_days_btw_first_last_review', 'int'),
 ('cancellation_policy', 'string'),
 ('price_c', 'string')]

In [0]:
data_model2=data_model2.dropna()

In [0]:
# Create a 70-30 train test split

train_data,test_data=data_model2.randomSplit([0.7,0.3])

In [0]:
# Import the required libraries

from pyspark.ml.classification import DecisionTreeClassifier
from pyspark.ml.feature import VectorAssembler,StringIndexer
from pyspark.ml import Pipeline


In [0]:
# Use StringIndexer to convert the categorical columns to hold numerical data

neighbourhood_indexer = StringIndexer(inputCol='neighbourhood',outputCol='neighbourhood_index',handleInvalid='keep')
property_type_indexer = StringIndexer(inputCol='property_type',outputCol='property_type_index',handleInvalid='keep')
room_type_indexer = StringIndexer(inputCol='room_type',outputCol='room_type_index',handleInvalid='keep')
bed_type_indexer = StringIndexer(inputCol='bed_type',outputCol='bed_type_index',handleInvalid='keep')
cancellation_policy_indexer = StringIndexer(inputCol='cancellation_policy',outputCol='cancellation_policy_index',handleInvalid='keep')
price_c_indexer = StringIndexer(inputCol='price_c',outputCol='price_c_index',handleInvalid='keep')


In [0]:
# Vector assembler is used to create a vector of input features

assembler = VectorAssembler(inputCols=['host_is_superhost','host_identity_verified','neighbourhood_index','property_type_index','room_type_index','accommodates','bathrooms','bedrooms','beds','bed_type_index','Number_of_amenities','guests_included','price_per_extra_person','minimum_nights','number_of_reviews','number_days_btw_first_last_review','cancellation_policy_index'],outputCol="features")

In [0]:
# Create an object for the Logistic Regression model
# Use the parameter maxBins and assign a value that is equal to or more than the number of categories in any sigle feature

dt_model = DecisionTreeClassifier(labelCol='price_c_index',maxBins=500)

In [0]:
# Pipeline is used to pass the data through indexer and assembler simultaneously. Also, it helps to pre-rocess the test data
# in the same way as that of the train data

pipe = Pipeline(stages=[neighbourhood_indexer,property_type_indexer,room_type_indexer,bed_type_indexer,cancellation_policy_indexer,
                        price_c_indexer,assembler,dt_model])

In [0]:
# It took 1.6 minutes for this step to execute

fit_model=pipe.fit(train_data)

In [0]:
# Store the results in a dataframe

results = fit_model.transform(test_data)

In [0]:
results.select(['price_c_index','prediction']).show()

+-------------+----------+
|price_c_index|prediction|
+-------------+----------+
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
+-------------+----------+
only showing top 20 rows



##### Model Evaluation - Decision Tree Classification
##### 1) Accuracy
##### 2) f1 score

In [0]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

In [0]:
ACC_evaluator = MulticlassClassificationEvaluator(
    labelCol="price_c_index", predictionCol="prediction", metricName="accuracy")

In [0]:
accuracy = ACC_evaluator.evaluate(results)

In [0]:
print("The accuracy of the decision tree classifier is {}".format(accuracy))

The accuracy of the decision tree classifier is 0.8365605551963984


In [0]:
f1_evaluator = MulticlassClassificationEvaluator(
    labelCol="price_c_index", predictionCol="prediction", metricName="f1")

In [0]:
f1_1 = f1_evaluator.evaluate(results)

In [0]:
print("The f1_Score of the decision tree classifier is {}".format(f1_1))

The f1_Score of the decision tree classifier is 0.8315319706697342


### Classification Model 2 - Random Forest Classification

In [0]:
from pyspark.ml.classification import RandomForestClassifier

In [0]:
# Train a rf model.
rf_model = RandomForestClassifier(labelCol="price_c_index", featuresCol="features", numTrees=10)

In [0]:
# Pipeline is used to pass the data through indexer and assembler simultaneously. Also, it helps to pre-rocess the test data
# in the same way as that of the train data

pipe = Pipeline(stages=[neighbourhood_indexer,property_type_indexer,room_type_indexer,bed_type_indexer,cancellation_policy_indexer,
                        price_c_indexer,assembler,rf_model])

In [0]:
# It took 1.6 minutes for this step to execute

fit_model=pipe.fit(train_data)

In [0]:
# Store the results in a dataframe

results_2 = fit_model.transform(test_data)

In [0]:
results_2.select(['price_c_index','prediction']).show()

+-------------+----------+
|price_c_index|prediction|
+-------------+----------+
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
+-------------+----------+
only showing top 20 rows



##### Model Evaluation - Random Forest Classification
##### 1) Accuracy
##### 2) f1 score

In [0]:
rf_acc_evaluator = MulticlassClassificationEvaluator(
    labelCol="price_c_index", predictionCol="prediction", metricName="accuracy")

In [0]:
accuracy_2 = rf_acc_evaluator.evaluate(results_2)

In [0]:
print("The accuracy of the random forest classifier is {}".format(accuracy_2))

The accuracy of the random forest classifier is 0.8470483771721437


In [0]:
f1_rf_evaluator = MulticlassClassificationEvaluator(
    labelCol="price_c_index", predictionCol="prediction", metricName="f1")

In [0]:
f1_rf = f1_rf_evaluator.evaluate(results_2)

In [0]:
print("The f1 score of the random forest is {}".format(f1_rf))

The f1 score of the random forest is 0.8314335581232121


### Classification Model 3 - MultilayerPerceptron Classification

In [0]:
from pyspark.ml.feature import OneHotEncoder, VectorAssembler, StringIndexer
from pyspark.ml import Pipeline
from pyspark.sql.functions import udf, StringType
from pyspark.ml.classification import MultilayerPerceptronClassifier
import pyspark.sql.functions


In [0]:
train, test = data_model2.randomSplit([0.7, 0.1], 1234)

In [0]:
categorical_columns = [item[0] for item in data_model2.dtypes if item[1].startswith(
    'string')]
int_columns = [item[0] for item in data_model2.dtypes if item[1].startswith('int')]
float_columns = [item[0] for item in data_model2.dtypes if item[1].startswith('float')]
indexers = [StringIndexer(inputCol=column, outputCol='{0}_index'.format(
    column)) for column in categorical_columns]

In [0]:
featuresCreator = VectorAssembler(
    inputCols=[indexer.getOutputCol() for indexer in indexers] + float_columns + int_columns,
    outputCol='features')
layers = [len(featuresCreator.getInputCols()), 4, 2, 3]

In [0]:
classifier = MultilayerPerceptronClassifier(labelCol='price_c_index',
                                            featuresCol='features',
                                            maxIter=100,
                                            layers=layers,
                                            blockSize=128,
                                            seed=1234)

In [0]:
pipeline = Pipeline(stages=indexers + [featuresCreator, classifier])
model = pipeline.fit(train)

In [0]:
test_output_df = model.transform(test)

In [0]:
test_output_df.select(['price_c_index','prediction']).show()

+-------------+----------+
|price_c_index|prediction|
+-------------+----------+
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
|          0.0|       0.0|
+-------------+----------+
only showing top 20 rows



##### Model Evaluation - Multilayer Perceptron Classification
##### 1) Accuracy
##### 2) f1 score

In [0]:
mpc_evaluator = MulticlassClassificationEvaluator(
    labelCol="price_c_index", predictionCol="prediction", metricName="accuracy")

In [0]:
accuracy_3 = mpc_evaluator.evaluate(test_output_df)

In [0]:
print("The accuracy of the Multilayer Perceptron classifier is {}".format(accuracy_3))

The accuracy of the Multilayer Perceptron classifier is 0.7491991744258805


In [0]:
f1_mpc_evaluator = MulticlassClassificationEvaluator(
    labelCol="price_c_index", predictionCol="prediction", metricName="f1")

In [0]:
f1_mpc = f1_mpc_evaluator.evaluate(test_output_df)

In [0]:
print("The f1 score of the Multilayer perceptron classifier is {}".format(f1_mpc))

The f1 score of the Multilayer perceptron classifier is 0.660557319280575


### Interpretation

The objective behind our project is to identify the locations around which the customers can find the most number of Airbnb listings and predicting the price category of Airbnb listings namely affordable, moderate, or expensive.

For identifying locations, we used K means clustering on latitude and longitude of Airbnb listings. From K means clustering, we were able to infer that:

29.6 % of the listings are centered around location with latitude 42.35114839 and longitude -71.13995305, which are the co-ordinates for 50 Gordon St, Boston, MA area.

28.6 % of the listings are centered around location with latitude 42.31555492and longitude -71.05618115, which are the co-ordinates for 68 Belfort St, Dorchester, Boston, MA area.
It can be suggested that tourists have higher chances of finding Airbnb listings for stay around the above suggested areas.

For predicting the price category of Airbnb listings into affordable (<$175), moderate (between $175 and $335) or expensive (>$335) categories, we have a multiclass classification problem. We used 17 attributes for our prediction models, such as if the host is super host, no. of bedrooms, neighborhood, cancellation policy, to name a few. The analysis was carried out using 3 classification models and below are the evaluation metrics for each model:

#### Decision Tree Classification:

Accuracy: 84%; F1 Score: 0.83

#### Random Forest Classification:

Accuracy: 85.2%; F1 Score: 0.84

#### Multi-Layer Perceptron Classification:
 
Accuracy: 74.9%; F1 Score: 0.66

Since our dataset used is imbalanced, we are not using Accuracy but F1 Score as our preferred evaluation metric for comparing the three models. F1 score provides a way to combine precision and recall metrices (harmonic mean of precision and recall). Precision is a metric that finds out what fraction of predicted positives is actually positive, i.e., out of the total predicted price categories, what fraction of them were correctly identified. Recall is a metric that quantifies the number of correct positive predictions (correctly identified prices) made of all positive predictions that could have been made.

Comparing all the 3 models, we could infer that Random Forest Classifier performs the best among the other models with F1 score of 0.84. Since the score is closer to 1, we conclude that
the model performs well in classifying the price of Airbnb listings into 3 categories- affordable, moderate, and expensive.